src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1846,57 +1848,91 @@ void ObjectSynchronizer::log_in_use_monitor_details(outputStream* out) {
   out->flush();
 }
 
-void SystemDictObjMonitor::lock(BasicLock* lock, TRAPS) {
+void SystemDictObjMonitor::lock(BasicLock* lock, Thread* current) {
   assert(UseWispMonitor, "SystemDictObjMonitor if only for UseWispMonitor");
+  if (!current->is_Java_thread()) {
+    _monitor->lock();
+    return;
+  }
+
+  JavaThread* jt = current->as_Java_thread();
   if (!is_obj_lock()) {
     _monitor->lock();
     if (is_obj_lock()) {
       // set_obj_lock has been called before we fetch the lock.
       // now _monitor is Pointless. re-fetch the objectMonitor to
       // satisfy the critical section semantics.
       _monitor->unlock();
-      ObjectSynchronizer::enter(Handle(THREAD, _obj), lock, THREAD);
+      ObjectSynchronizer::enter(Handle(jt, _obj.resolve()), lock, jt);
     }
   } else {
-    ObjectSynchronizer::enter(Handle(THREAD, _obj), lock, THREAD);
+    ObjectSynchronizer::enter(Handle(jt, _obj.resolve()), lock, jt);
   }
 }
 
-void SystemDictObjMonitor::unlock(BasicLock* lock, TRAPS) {
+void SystemDictObjMonitor::unlock(BasicLock* lock, Thread* current) {
   assert(UseWispMonitor, "SystemDictObjMonitor if only for UseWispMonitor");
+  if (!current->is_Java_thread()) {
+    _monitor->unlock();
+    return;
+  }
+
+  JavaThread* jt = current->as_Java_thread();
   if (!is_obj_lock()) {
     _monitor->unlock();
   } else {
-    ObjectSynchronizer::exit(_obj, lock, THREAD);
+    ObjectSynchronizer::exit(_obj.resolve(), lock, jt);
   }
 }
 
-void SystemDictObjMonitor::wait(BasicLock* lock, TRAPS) {
+void SystemDictObjMonitor::wait(BasicLock* lock, Thread* current) {
   assert(UseWispMonitor, "SystemDictObjMonitor if only for UseWispMonitor");
+  if (!current->is_Java_thread()) {
```

Comment:
There is no check for `is_obj_lock`, if another thread calls `ObjectSynchronizer::notifyall`, the current thread will not be woken up

Reply:
Firstly, we check if the thread is `JavaThread`. If current thread is not a `JavaThread`, there is no need to use `ObjectSynchronizer`. And then, if current thread is JavaThread we will check `is_obj_lock` to do `SystemDictObjMonitor::wait`.